dot/sync/chain_sync.go

```diff
@@ -0,0 +1,785 @@
+	return nil
+}
+
+func (cs *chainSync) logSyncSpeed() {
+	t := time.NewTicker(time.Second * 5)
+	defer t.Stop()
+
+	for {
+		before, err := cs.blockState.BestBlockHeader()
+		if err != nil {
+			continue
+		}
+
+		if cs.state == bootstrap {
+			cs.benchmarker.begin(before.Number.Uint64())
+		}
+
+		select {
+		case <-t.C:
+			// TODO: why does this function not return when ctx is cancelled???
+			if cs.ctx.Err() != nil {
+				return
+			}
+		case <-cs.ctx.Done():
+			return
+		}
+
+		finalised, err := cs.blockState.GetHighestFinalisedHeader()
+		if err != nil {
+			continue
+		}
+
+		switch cs.state {
+		case bootstrap:
+			after, err := cs.blockState.BestBlockHeader()
+			if err != nil {
+				continue
+			}
+
+			cs.benchmarker.end(after.Number.Uint64())
+			target := cs.getTarget()
+
+			logger.Info("🔗 imported blocks", "from", before.Number, "to", after.Number,
+				"hashes", fmt.Sprintf("[%s ... %s]", before.Hash(), after.Hash()),
+			)
+
+			logger.Info("🚣 currently syncing",
+				"peer count", len(cs.network.Peers()),
+				"target", target,
+				"average blocks/second", cs.benchmarker.mostRecentAverage(),
+				"overall average", cs.benchmarker.average(),
+				"finalised", finalised.Number,
+				"hash", finalised.Hash(),
+			)
+		case tip:
+			logger.Info("💤 node waiting",
+				"peer count", len(cs.network.Peers()),
+				"head", before.Number,
+				"hash", before.Hash(),
+				"finalised", finalised.Number,
+				"hash", finalised.Hash(),
+			)
+		}
+	}
+}
+
+func (cs *chainSync) sync() {
+	// set to slot time * 2
+	// TODO: make configurable
+	ticker := time.NewTicker(time.Second * 12)
+
+	for {
+		select {
+		case ps := <-cs.workQueue:
+			head, err := cs.blockState.BestBlockHeader()
+			if err != nil {
+				logger.Error("failed to get best block header", "error", err)
+				continue
+			}
+
+			target := cs.getTarget()
+			if big.NewInt(0).Add(head.Number, big.NewInt(maxResponseSize)).Cmp(target) == -1 {
+				// we are 128 blocks or more behind the target, switch to bootstrap mode
+				logger.Debug("switching to bootstrap sync mode...")
+				cs.setMode(bootstrap)
+			} else {
+				// bootstrap complete, switch state to tip if not already
+				// and begin near-head fork-sync
+				logger.Debug("switching to tip sync mode...")
+				cs.setMode(tip)
+			}
+
+			if err := cs.handleWork(ps); err != nil {
+				logger.Error("failed to handle chain sync work", "error", err)
+			}
+		case res := <-cs.resultQueue:
+			// delete worker from workers map
+			cs.workerState.delete(res.id)
+
+			// handle results from worker
+			// if there is an error, potentially retry the worker
+			if res.err == nil {
+				// TODO: log worker time
+				continue
+			}
+
+			logger.Error("worker error", "error", res.err.err)
+
+			// handle errors. in the case that a peer did not respond to us in time,
+			// temporarily add them to the ignore list.
+			// TODO: periodically clear out ignore list, currently is done if (ignore list >= peer list)
+			switch res.err.err {
```

Comment:
Switch on an error will fail if errors are wrapped errors. I think we are not wrapping any errors as of now. We can add a note here or use `error.Is`

Reply:
I'll update it to the `errors.Is`